BLE_Privacy/source/main.cpp

```diff
@@ -0,0 +1,575 @@
+         * using our event queue */
+        _ble.onEventsToProcess(
+            makeFunctionPointer(this, &PrivacyDevice::schedule_ble_events)
+        );
+
+        error = _ble.init(this, &PrivacyDevice::on_init_complete);
+
+        if (error) {
+            printf("Error returned by BLE::init.\r\n");
+            return;
+        }
+
+        /* this will not return until shutdown */
+        _event_queue.dispatch_forever();
+    };
+
+    /** Override to start chosen activity when initialisation completes */
+    virtual void start() = 0;
+
+    /** Override to start chosen activity after initial bonding */
+    virtual void start_after_bonding() = 0;
+
+    /* event handler functions */
+
+    /** Inform the application of pairing. When succesful we will attempt
+     *  to create a whitelist based on the newly created bond */
+    virtual void pairingResult(
+        ble::connection_handle_t connectionHandle,
+        SecurityManager::SecurityCompletionStatus_t result
+    ) {
+        if (result == SecurityManager::SEC_STATUS_SUCCESS) {
+            printf("Pairing successful\r\n");
+
+            if (!_bonded) {
+                /* generate whitelist only the first time we bond */
+                _bonded = true;
+
+                /* provide memory for the generated whitelist */
+                Gap::Whitelist_t* whitelist = new Gap::Whitelist_t;
+                whitelist->size = 0;
+                /* we only need space for one address in our demonstration
+                 * but this can return all the bonded addresses */
+                whitelist->capacity = 1;
+                whitelist->addresses = new BLEProtocol::Address_t();
+
+                /* this will only fill the whitelist up to the provided capacity,
+                 * we hand over the memory ownership to the function */
+                ble_error_t error = _ble.securityManager().generateWhitelistFromBondTable(whitelist);
+
+                if (error != BLE_ERROR_NONE) {
+                    delete whitelist->addresses;
+                    delete whitelist;
+                }
+            }
+        } else {
+            printf("Pairing failed\r\n");
+        }
+
+        /* disconnect in 1000 ms */
+        _event_queue.call_in(
+            1000, &_ble.gap(),
+            &Gap::disconnect, _handle, Gap::REMOTE_USER_TERMINATED_CONNECTION
+        );
+    }
+
+
+    virtual void whitelistFromBondTable(Gap::Whitelist_t* whitelist)
+    {
+        if (whitelist->size) {
+            printf("Whitelist generated.\r\n");
+            _whitelist_generated = true;
+
+            /* set the newly created whitelist at the link layer,
+             * see BLUETOOTH SPECIFICATION Version 5.0 | Vol 6, Part B - 4.3 */
+            _ble.gap().setWhitelist(*whitelist);
+        } else {
+            printf("Whitelist failed to generate.\r\n");
+        }
+
+        /* this callback transfer memory ownership to us so we have to dispose of it */
```

Comment:
Actually we have ownership of it during the whole process.

Reply:
we are not allowed to delete it before the callback happens and we do not store a pointer to it so effectively we are handing it over

Reply:
I will now seemingly backtrack on what I said: lifetime is decided by the user, the whitelist could be statically allocated, in my example I relinquish control but the doc doesn't require that: the only requirement is that you do not access the whitelist when callback is pending.